File: Rust/src/0201 ~ 0300/0223 - Rectangle Area.rs
// 链接：https://leetcode.com/problems/rectangle-area/
// 题意：给定两个矩形，求它们形成的图形的总面积？

// 输入： A = -3, B = 0, C = 3, D = 4, E = 0, F = -1, G = 9, H = 2
// 输出： 45

// 思路： 枚举
//
//      首先保证第一个矩形的左侧在第二个矩形的左侧左边即可
//      先求两个矩形的面积和 area_sum ，
//		枚举判断所有情况即可：
//      1. 两个矩形不相交，直接返回面积和 area_sum
//      2. 两个矩形相交，相交矩形的面积为 area ，则总面积为 area_sum - area
//          (1) 相交矩形的左侧 = 第二个矩形的左侧
//          (2) 相交矩形的下侧 = max(第一个矩形的下侧，第二个矩形的下侧)
//          (3) 相交矩形的右侧 = min(第一个矩形的右侧，第二个矩形的右侧)
//          (4) 相交矩形的上侧 = min(第一个矩形的上侧，第二个矩形的上侧)
//
//      时间复杂度： O(height ^ 2)
//      空间复杂度： O(height) 【栈的开销】

use std::cmp;

impl Solution {
    pub fn compute_area(a: i32, b: i32, c: i32, d: i32, e: i32, f: i32, g: i32, h: i32) -> i32 {
        // 如果第一个矩形的左侧在第二个矩形的的左侧右边，则调整两个矩形的位置
        if a > e {
            return Solution::compute_area(e, f, g, h, a, b, c, d);
        }
        // 现在第一个矩形的左侧必定在第二个矩形的左侧左边
        // 算出两个矩形的面积和，等下减去相交矩形的面积即可
        let area_sum = Solution::area(a, b, c, d) + Solution::area(e, f, g, h);

        // 如果第二个矩形右侧在第一个矩形右侧右边 或者
        //    第二个矩形下侧在第一个矩形上侧上边 或者
        //    第二个矩形上侧在第一个矩形下侧下边
        // 则两个矩形不相交，直接返回面积和即可
        if e > c || f > d || h < b {
            return area_sum;
        }

        // 现在两个矩形必定相交
        // 相交矩形的左侧 = 第二个矩形的左侧
        let i = e;
        // 相交矩形的下侧 = max(第一个矩形的下侧，第二个矩形的下侧)
        let j = cmp::max(b, f);
        // 相交矩形的右侧 = min(第一个矩形的右侧，第二个矩形的右侧)
        let k = cmp::min(c, g);
        // 相交矩形的上侧 = min(第一个矩形的上侧，第二个矩形的上侧)
        let l = cmp::min(d, h);

        return area_sum - Solution::area(i, j, k, l)
    }

    pub fn area(a: i32, b: i32, c: i32, d: i32) -> i32 {
        (c - a) * (d - b)
    }
}

File: Go/src/0201 ~ 0300/0223 - Rectangle Area.go
// 链接：https://leetcode.com/problems/rectangle-area/
// 题意：给定两个矩形的坐标，求它们形成的图形的总面积？
//      第一个矩形的左下角为 (ax1, ay1) ，右上角为 (ax2, ay2) 。
//      第一个矩形的左下角为 (bx1, by1) ，右上角为 (bx2, by2) 。


// 数据限制：
//  -10 ^ 4 <= ax1 <= ax2 <= 10 ^ 4
//  -10 ^ 4 <= ay1 <= ay2 <= 10 ^ 4
//  -10 ^ 4 <= bx1 <= bx2 <= 10 ^ 4
//  -10 ^ 4 <= by1 <= by2 <= 10 ^ 4


// 输入： ax1 = -3, ay1 = 0, ax2 = 3, ay2 = 4, bx1 = 0, by1 = -1, bx2 = 9, by2 = 2
// 输出： 45

// 输入： ax1 = -2, ay1 = -2, ax2 = 2, ay2 = 2, bx1 = -2, by1 = -2, bx2 = 2, by2 = 2
// 输出： 16


// 思路： 模拟
//
//      题目 LeetCode 2446 的加强版，需要求出相交的长度。
//
//      对于 x 轴上的两个线段 (ax1, ax2) 和 (bx1, bx2) ，
//      当且仅当 ax1 <= bx2 && bx1 <= ax2 时，两个线段相交。
//
//      相交线段的左端点 left = max(ax1, bx1) ，右端点 right = min(ax2, bx2) ，
//      那么相交线段长度为 right - left 。
//
//      同理，对于 y 轴上的两个线段 (ay1, ay2) 和 (by1, by2) ，
//      当且仅当 ay1 <= by2 && by1 <= ay2 时，两个线段相交。
//
//      相交线段的下端点 bottom = max(ay1, by1) ，上端点 top = min(ay2, by2) ，
//      那么相交线段长度为 top - bottom 。
//
//      那么当且仅当两个矩形在 x 方向和 y 方向都相交时，它们存在重叠区域，
//      重叠区域的面积为 (right - left) * (top - bottom) 。
//
//      我们可以用做差法得计算公式：总面积 = 两个矩形的面积和 - 两个矩形重叠区域的面积。
//
//
//      时间复杂度：O(1)
//          1. 只需要常数次布尔运算和算数运算即可
//      空间复杂度：O(1)
//          1. 只需要维护常数个额外变量


func computeArea(ax1 int, ay1 int, ax2 int, ay2 int, bx1 int, by1 int, bx2 int, by2 int) int {
    // x 表示矩形 x 方向相交长度。初始化为 0 ，表示不相交，方便后续处理
    x := 0
    // 如果 x 方向相交，则更新相交长度
    if ax1 <= bx2 && bx1 <= ax2 {
        x = min(ax2, bx2) - max(ax1, bx1)
    }

    // y 表示矩形 y 方向相交长度。初始化为 0 ，表示不相交，方便后续处理
    y := 0
    // 如果 y 方向相交，则更新相交长度
    if ay1 <= by2 && by1 <= ay2 {
        y = min(ay2, by2) - max(ay1, by1)
    }

    // 总面积 = 两个矩形的面积和 - 两个矩形重叠区域的面积
    return area(ax1, ay1, ax2, ay2) + area(bx1, by1, bx2, by2) - x * y
}

// 计算矩形（左下角为 (x1, y1) ，右上角为 (x2, y2) ）的面积
func area(x1, y1, x2, y2 int) int {
    return (x2 - x1) * (y2 - y1)
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
